src/app/reporting/Read-Checked.cpp

```diff
@@ -0,0 +1,132 @@
+ *    limitations under the License.
+ */
+#include <app/reporting/Read-Checked.h>
+
+#include <app/reporting/Read-DataModel.h>
+#include <app/reporting/Read-Ember.h>
+#include <app/util/MatterCallbacks.h>
+
+namespace chip {
+namespace app {
+namespace reporting {
+namespace CheckedImpl {
+namespace {
+
+/// Checkpoints and saves the state (including error state) for a
+/// AttributeReportIBs::Builder
+class ScopedAttributeReportIBsBuilderState
+{
+public:
+    ScopedAttributeReportIBsBuilderState(AttributeReportIBs::Builder & builder) : mBuilder(builder), mError(mBuilder.GetError())
+    {
+        mBuilder.Checkpoint(mCheckpoint);
+    }
+
+    ~ScopedAttributeReportIBsBuilderState()
+    {
+        mBuilder.Rollback(mCheckpoint);
+        mBuilder.ResetError(mError);
+    }
+
+private:
+    AttributeReportIBs::Builder & mBuilder;
+    chip::TLV::TLVWriter mCheckpoint;
+    CHIP_ERROR mError;
+};
+
+} // namespace
+
+CHIP_ERROR RetrieveClusterData(InteractionModel::DataModel * dataModel, const Access::SubjectDescriptor & subjectDescriptor,
+                               bool isFabricFiltered, AttributeReportIBs::Builder & reportBuilder,
+                               const ConcreteReadAttributePath & path, AttributeEncodeState * encoderState)
+{
+    ChipLogDetail(DataManagement, "<RE:Run> Cluster %" PRIx32 ", Attribute %" PRIx32 " is dirty", path.mClusterId,
+                  path.mAttributeId);
+    DataModelCallbacks::GetInstance()->AttributeOperation(DataModelCallbacks::OperationType::Read,
+                                                          DataModelCallbacks::OperationOrder::Pre, path);
+
+    AttributeEncodeState stateDm(encoderState); // a copy for DM logic only. Ember changes state directly
+
+    CHIP_ERROR errEmber         = CHIP_NO_ERROR;
+    uint32_t lengthWrittenEmber = 0;
+
+    {
+        ScopedAttributeReportIBsBuilderState builderState(reportBuilder); // temporary only
+        errEmber =
+            EmberImpl::RetrieveClusterData(dataModel, subjectDescriptor, isFabricFiltered, reportBuilder, path, encoderState);
+        lengthWrittenEmber = reportBuilder.GetWriter()->GetLengthWritten();
+    }
+
+    CHIP_ERROR errDM =
+        DataModelImpl::RetrieveClusterData(dataModel, subjectDescriptor, isFabricFiltered, reportBuilder, path, &stateDm);
+
+    if (errEmber != errDM)
+    {
+        // Note log + chipDie instead of VerifyOrDie so that breakpoints (and usage of rr)
+        // is easier to debug.
+        ChipLogError(Test, "Different return codes between ember and DM");
+        ChipLogError(Test, "  Ember error: %" CHIP_ERROR_FORMAT, errEmber.Format());
+        ChipLogError(Test, "  DM error: %" CHIP_ERROR_FORMAT, errDM.Format());
+        chipDie();
+    }
+
+    // data should be identical for most cases EXCEPT that for time-deltas (e.g. seconds since boot or similar)
+    // it may actually differ. As a result, the amount of data written in bytes MUST be the same, however if the rest of the
```

Comment:
Technicaly not always 100% accurate: if data size changes, e.g. a stamp changing from 0xFF to 0x100 the sizes of written data will also differ.

I am unsure if we can actually detect this reliably or if it is that likely. We will use checked for internal tests only (otherwise it is both slow and bloaty), so for now leaning towards leaving it as-is.

Reply:
I updated the comment and logic.